loadgen/cmd/otelbench/main.go

```diff
@@ -0,0 +1,72 @@
+// Licensed to Elasticsearch B.V. under one or more contributor
+// license agreements. See the NOTICE file distributed with
+// this work for additional information regarding copyright
+// ownership. Elasticsearch B.V. licenses this file to you under
+// the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing,
+// software distributed under the License is distributed on an
+// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+
+package main
+
+import (
+	"context"
+	"flag"
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/elastic/opentelemetry-collector-components/loadgen"
+)
+
+func main() {
+	loadgen.Init()
+	flag.CommandLine = loadgen.FlagSet
+	testing.Init()
+	flag.Parse()
+
+	result := testing.Benchmark(func(b *testing.B) {
+		stop := make(chan bool)
+
+		go func() {
+			ticker := time.NewTicker(500 * time.Millisecond)
+			defer ticker.Stop()
+			for {
+				select {
+				case <-stop:
+				case <-ticker.C:
+					logs, metricPoints, spans, err := loadgen.GetTelemetrySent()
+					if err != nil {
+						b.Logf("error getting internal telemetry: %s", err)
+						continue
+					}
+					total := logs + metricPoints + spans
```

Comment:
This feels a bit wrong. `b.N` is meant to count iterations, and in this case I would expect one iteration to be sending all the configured payloads once, so each iteration is equivalent. The asynchronous check also makes me a bit uncomfortable -- it may be a bit inaccurate as a result.

Perhaps the `loadgenreceiver.NewFactory` function could take some options, one of which is a channel for controlling the iterations of the receiver? Then the otelbench command could inject a channel. We shouldn't need another goroutine then: the `testing.B` loop would be interacting with the receiver through the channel.

Reply:
I'm wondering how to do this at implementation level, since logs, metrics and traces pipelines are separate. Do we want logs + metrics + traces iterations to be greater than b.N? My PR at its existing state has the same problem, which causes some imbalance between log records, metric data points and spans. Should we just benchmark the signals separately?

Reply:
> Should we just benchmark the signals separately?

That sounds like a good idea. They could have reasonably different performance profiles too, so combining them may be unhelpful.

Reply:
Injecting channels to NewFactory for testing.Benchmark to get notified when collector finishes sending b.N iterations of data set.